### PR TITLE
plugin update. Closes #164 Closes #176 Closes #177

### LINKED ISSUE #164: check for latest plugin versions
check for new versions of plugins which were installed with the `latest` tag

### LINKED ISSUE #176: implement `--update-all` flag in `plugin install`

### LINKED ISSUE #177: fix issue where config digest is recorded in `versions.json`

### PR DESCRIPTION
**Pull request by binaek89**

### CODE REVIEW

**Comment by binaek89** [comment]
Current plugins:
```
binaeksarkar@Binaeks-MacBook-Pro ~ % steampipe plugin list           
┌──────────────────────────────────────────────┬─────────┬─────────────┐
│ NAME                                         │ VERSION │ CONNECTIONS │
├──────────────────────────────────────────────┼─────────┼─────────────┤
│ hub.steampipe.io/plugins/turbot/aws@latest   │ 0.4.0   │ aws2        │
│ hub.steampipe.io/plugins/turbot/chaos@latest │ 0.0.4   │ chaos       │
└──────────────────────────────────────────────┴─────────┴─────────────┘

```

```
binaeksarkar@Binaeks-MacBook-Pro ~ % steampipe plugin install aws gcp

Error: install failed for plugin 'aws': aws is already installed. If you want update it, please run steampipe plugin update aws

Installed plugin: gcp v0.0.6
Documentation:    https://hub.steampipe.io/plugins/turbot/gcp

```
```
binaeksarkar@Binaeks-MacBook-Pro ~ % steampipe plugin list
┌──────────────────────────────────────────────┬─────────┬─────────────┐
│ NAME                                         │ VERSION │ CONNECTIONS │
├──────────────────────────────────────────────┼─────────┼─────────────┤
│ hub.steampipe.io/plugins/turbot/aws@latest   │ 0.4.0   │ aws2        │
│ hub.steampipe.io/plugins/turbot/chaos@latest │ 0.0.4   │ chaos       │
│ hub.steampipe.io/plugins/turbot/gcp@latest   │ 0.0.6   │ gcp         │
└──────────────────────────────────────────────┴─────────┴─────────────┘
```
```
binaeksarkar@Binaeks-MacBook-Pro ~ % steampipe plugin update azure aws gcp

Error: update failed for plugin 'azure': azure is not installed. If you want install it, please run steampipe plugin install azure

Updated plugin: aws v0.4.0
Documentation:    https://hub.steampipe.io/plugins/turbot/aws

Updated plugin: gcp v0.0.6
Documentation:    https://hub.steampipe.io/plugins/turbot/gcp
```
```
binaeksarkar@Binaeks-MacBook-Pro ~ % steampipe plugin install aws gcp     

Error: install failed for plugin 'aws': aws is already installed. If you want update it, please run steampipe plugin update aws

Error: install failed for plugin 'gcp': gcp is already installed. If you want update it, please run steampipe plugin update gcp
```
```
binaeksarkar@Binaeks-MacBook-Pro ~ % steampipe plugin update aws gcp 

Updated plugin: aws v0.4.0
Documentation:    https://hub.steampipe.io/plugins/turbot/aws

Updated plugin: gcp v0.0.6
Documentation:    https://hub.steampipe.io/plugins/turbot/gcp

```

File: pluginmanager/pluginupdate.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+		Org:     org,
+		Name:    name,
+		Stream:  stream,
+		Version: plugin.Version,
+		Digest:  plugin.ImageDigest,
+	}
+	return payload
+}
+
+func (pvc *PluginVersionChecker) splitNameIntoOrgNameAndStream(name string) (string, string, string) {
```
Comment (kaidaguerre) [inline pluginmanager/pluginupdate.go:132]: is there not existing functionality to do this kind of manipulation?

File: task/pluginversioncheck.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"strings"
+
+	"github.com/fatih/color"
+	"github.com/olekukonko/tablewriter"
+	"github.com/turbot/steampipe/pluginmanager"
+)
+
+// check if there is a new version
+func checkPluginVersions(installationID string) {
+	// if SP_DISABLE_UPDATE_CHECK is set, do nothing
```
Comment (kaidaguerre) [inline task/pluginversioncheck.go:15]: need to handle old and new vars - see main

File: task/runner.go [hunk 3]
```diff
@@ -1,70 +1,58 @@
...
-	r.loadState()
+	r.currentState, _ = statefile.LoadState()
 	r.shouldRun = r.getShouldRun()
 	return r
 }
 
 func (r *Runner) Run() {
 	if r.shouldRun {
+		waitGroup := sync.WaitGroup{}
+		waitGroup.Add(3)
```
Comment (kaidaguerre) [inline task/runner.go:28]: I'd rather have a separate `waitGroup.Add(1)` begore each go routine - that way if we add another we will hopefully remember to inc the wg

File: task/runner.go [hunk 4]
```diff
@@ -1,70 +1,58 @@
...
-		return
-	}
+func (r *Runner) runAsyncJob(job func(), wg *sync.WaitGroup) {
+	job()
+	wg.Done()
 }
 
 func (r *Runner) getShouldRun() bool {
 	now := time.Now()
+	if r.currentState == nil {
```
Comment (kaidaguerre) [inline task/runner.go:53]: is this right - if there is no current state shouldn't we run for the first time?

File: task/taskutils.go [hunk 5]
```diff
@@ -0,0 +1,3 @@
+package task
+
```
Comment (kaidaguerre) [inline task/taskutils.go:2]: hmm I think you need to rebase from main, this is defined in `task/version_checker.go`, with legacy and new version

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

File: task/version_checker.go [hunk 6]
```diff
@@ -41,23 +42,17 @@ type versionCheckRequest struct {
...
-	signature    string // flags whether update check should be done
+	checkResult *versionCheckResponse // a channel to store the HTTP response
+	disabled    bool
+	signature   string // flags whether update check should be done
 }
 
 // check if there is a new version
-func checkVersion(id string) {
+func checkSteampipeVersion(id string) {
+	// if SP_DISABLE_UPDATE_CHECK is set, do nothing
```
Comment (kaidaguerre) [inline task/version_checker.go:55]: update comment to use new env var

**Review by kaidaguerre — CHANGES_REQUESTED** [review]

**Pull request merged**